### ISSUE TITLE
Cas-attributes are not stored as headers if CASAuthNHeader is unset

### ISSUE THREAD

**Issue by mostolog** [Jasig/mod_auth_cas#101]
Hi

As discussed in https://groups.google.com/a/apereo.org/d/msg/cas-user/_UASHFvNYi4/vduQUbmmAQAJ and according to documentation:
> Directive:	CASAuthNHeader
> Default:	None
> Description:	If enabled, this will store the user returned by CAS in an HTTP header
> 		accessible to your web applications.

Although this is true, seems incomplete because **SAML attributes** (those specified with _Require cas-attribute_/returned by CAS) **aren't stored as headers if CASAuthNHeader is not set**.

**Please, update documentation** to something like:
 "If enabled, this will store the CAS returned attributes, including the logged user in a header named by CASAuthNHeader directive, as HTTP headers accessible to your web applications."

Also, have in mind that _CASScrubRequestHeaders_ strips headers, including _CASAuthNHeader_. In order to store cas-attributes as headers, something like this will do the trick:

_CASAuthNHeader foo_ will add a _foo_ header naming logged user
_CASAttributePrefix foo_ will prefix cas-attributes with _foo_
_CASScrubRequestHeaders On_ will strip all cas-attributes from headers

Hence, we need something like _CASAttributesAsHeaders On_ by default, having backward compatibility.

Regards

**Comment by dhawes** [comment]
The documentation update is simple enough and would be useful in the short term.

I'm debating whether it makes sense to add another directive that controls whether the SAML attributes will be added the the request headers. Maybe something like CASAuthzHeaders?

**Comment by dhawes** [comment]
> Also, have in mind that CASScrubRequestHeaders strips headers, including CASAuthNHeader.

This is false. Only inbound request headers are scrubbed. Those set by mod_auth_cas are retained.

**Comment by adaugherity** [comment]
I'm working on a PR to improve the docs for attributes, as most of it refers to "SAML attributes", but after #110 was merged, they may just be CAS 2.0/3.0 attributes, not SAML.  (Also, that PR is logging the attributes at `APLOG_ERR`, which at first led me to believe there was an error parsing them; I changed it to the more appropriate `APLOG_DEBUG`.)

After struggling to get attributes seen by a CGI script, I went looking through the source code and discovered this very issue (which I will also document).

However, even after setting `CASAuthNHeader`, I was _still_ not seeing any other attributes in CGI.  I eventually found this issue, and most importantly, in the linked Google group discussion:

> Also note that you may want to set your CASAttributePrefix to 
something that does not contain underscores, as Apache 2.4 will 
silently drop those headers. 

Wow, talk about burying the lede -- this was in fact my problem!  After setting `CASAttributePrefix "X-CAS-"` everything started working!

Based on this, wouldn't it be wise to change the default `CASAttributePrefix` (at least when building under Apache 2.4+) to something that isn't automatically dropped?

I do agree with gating setting HTTP headers for CAS attributes on a new boolean config directive instead of whether `CASAuthNHeader` is set or not.

**Comment by dhawes** [comment]
#102 is on my short term list to get in place. These issues have languished for too long, and I hope to knock them out in the next few weeks.

A PR for documentation would be great.
